### ISSUE TITLE
Arrow dictionaries -> Polars Categorical

### ISSUE THREAD

**Issue by kdkavanagh** [pola-rs/polars#1308]
From reading through historical issues, I can reason through why Polars would convert arrow dictionaries to <something else>, but is there anything which prevents us from reading arrow dictionaries directly into a polars categorical instead of a large-utf8 dtype as indicated here (#725)?  

I ask because this conversion to strings has the possibility to massively inflate memory requirements depending on the column's cardinality and the length of the dataframe.  Happy to take a stab at this if you can point me in the right direction.

**Comment by ritchie46** [comment]
Yes, we could convert the arrow dictionaries directly to Categorical. I agree that would be better. 

Conversion is done [here](https://github.com/pola-rs/polars/blob/a5e72b20a0b09048a8ee1cc5dad46e7255a80d0f/polars/polars-core/src/series/mod.rs#L1815) and the dictionary type should be added there.

**Comment by ritchie46** [comment]
I made round trip of Utf8 to Categorical possible in #1380

**Comment by ghuls** [comment]
@ritchie46 

It doesn't seem to work (polars 0.9.9):

```python
In [67]: df_pl
Out[67]: 
shape: (102753538, 6)
┌──────────────┬─────────┬─────────┬──────────────────────┬───────────────┬──────────────────────┐
│ Chromosome   ┆ Start   ┆ End     ┆ CellBarcode          ┆ FragmentCount ┆ hash                 │
│ ---          ┆ ---     ┆ ---     ┆ ---                  ┆ ---           ┆ ---                  │
│ cat          ┆ i32     ┆ i32     ┆ cat                  ┆ i32           ┆ u64                  │
╞══════════════╪═════════╪═════════╪══════════════════════╪═══════════════╪══════════════════════╡
│ "chr1"       ┆ 3000052 ┆ 3000361 ┆ "GGACATAAGGGCCACT-1" ┆ 1             ┆ 4276748679109407223  │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "chr1"       ┆ 3000092 ┆ 3000324 ┆ "CGCGATTCAGGAACAT-1" ┆ 1             ┆ 15578130322035353946 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "chr1"       ┆ 3000132 ┆ 3000184 ┆ "ATATGGTGTTAGACCA-1" ┆ 1             ┆ 1638472252710708676  │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "chr1"       ┆ 3000134 ┆ 3000251 ┆ "ACCTTCATCTTTGAGA-1" ┆ 1             ┆ 11591604927233528336 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ ...          ┆ ...     ┆ ...     ┆ ...                  ┆ ...           ┆ ...                  │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1946    ┆ 1976    ┆ "TCCTTGCAGCTCAATA-1" ┆ 1             ┆ 3797976091205386816  │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1946    ┆ 1976    ┆ "TTCCACGGTTAACACG-1" ┆ 1             ┆ 13208961099853080992 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1955    ┆ 1976    ┆ "CTGTATTTCCGCTAGA-1" ┆ 2             ┆ 1491265187212498152  │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1955    ┆ 1976    ┆ "CTTTAGTTCCACCCTG-1" ┆ 3             ┆ 5441796412465218466  │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1955    ┆ 1976    ┆ "GACCTAGTCCTCAGCT-1" ┆ 1             ┆ 4242163593946851932  │
└──────────────┴─────────┴─────────┴──────────────────────┴───────────────┴──────────────────────┘

In [68]: %time df_arrow = df_pl.to_arrow()
CPU times: user 8.87 s, sys: 3.07 s, total: 11.9 s
Wall time: 11.9 s

In [69]: df_arrow
Out[69]: 
pyarrow.Table
Chromosome: large_string
Start: int32
End: int32
CellBarcode: large_string
FragmentCount: int32
hash: uint64
```

**Comment by ghuls** [comment]
Now it works fine.

polars categorical to arrow dict is very fast now (zero copy I assume).

arrow dict to polars categorical requires some time (but works).

```python
In [13]: df_pl
Out[13]: 
shape: (102753538, 5)
┌──────────────┬─────────┬─────────┬──────────────────────┬───────────────┐
│ Chromosome   ┆ Start   ┆ End     ┆ CellBarcode          ┆ FragmentCount │
│ ---          ┆ ---     ┆ ---     ┆ ---                  ┆ ---           │
│ cat          ┆ i32     ┆ i32     ┆ cat                  ┆ i32           │
╞══════════════╪═════════╪═════════╪══════════════════════╪═══════════════╡
│ "chr1"       ┆ 3000052 ┆ 3000361 ┆ "GGACATAAGGGCCACT-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "chr1"       ┆ 3000092 ┆ 3000324 ┆ "CGCGATTCAGGAACAT-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "chr1"       ┆ 3000132 ┆ 3000184 ┆ "ATATGGTGTTAGACCA-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "chr1"       ┆ 3000134 ┆ 3000251 ┆ "ACCTTCATCTTTGAGA-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ ...          ┆ ...     ┆ ...     ┆ ...                  ┆ ...           │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1946    ┆ 1976    ┆ "TCCTTGCAGCTCAATA-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1946    ┆ 1976    ┆ "TTCCACGGTTAACACG-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1955    ┆ 1976    ┆ "CTGTATTTCCGCTAGA-1" ┆ 2             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1955    ┆ 1976    ┆ "CTTTAGTTCCACCCTG-1" ┆ 3             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1955    ┆ 1976    ┆ "GACCTAGTCCTCAGCT-1" ┆ 1             │
└──────────────┴─────────┴─────────┴──────────────────────┴───────────────┘

In [14]: %time df_arrow = df_pl.to_arrow()
CPU times: user 365 µs, sys: 137 µs, total: 502 µs
Wall time: 389 µs

In [15]: df_arrow
Out[15]: 
pyarrow.Table
Chromosome: dictionary<values=large_string, indices=uint32, ordered=0>
Start: int32
End: int32
CellBarcode: dictionary<values=large_string, indices=uint32, ordered=0>
FragmentCount: int32

In [16]: %time df_pl_from_arrow = pl.from_arrow(df_arrow)
CPU times: user 6.11 s, sys: 1.09 s, total: 7.2 s
Wall time: 7.17 s

In [17]: df_pl_from_arrow
Out[17]: 
shape: (102753538, 5)
┌──────────────┬─────────┬─────────┬──────────────────────┬───────────────┐
│ Chromosome   ┆ Start   ┆ End     ┆ CellBarcode          ┆ FragmentCount │
│ ---          ┆ ---     ┆ ---     ┆ ---                  ┆ ---           │
│ cat          ┆ i32     ┆ i32     ┆ cat                  ┆ i32           │
╞══════════════╪═════════╪═════════╪══════════════════════╪═══════════════╡
│ "chr1"       ┆ 3000052 ┆ 3000361 ┆ "GGACATAAGGGCCACT-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "chr1"       ┆ 3000092 ┆ 3000324 ┆ "CGCGATTCAGGAACAT-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "chr1"       ┆ 3000132 ┆ 3000184 ┆ "ATATGGTGTTAGACCA-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "chr1"       ┆ 3000134 ┆ 3000251 ┆ "ACCTTCATCTTTGAGA-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ ...          ┆ ...     ┆ ...     ┆ ...                  ┆ ...           │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1946    ┆ 1976    ┆ "TCCTTGCAGCTCAATA-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1946    ┆ 1976    ┆ "TTCCACGGTTAACACG-1" ┆ 1             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1955    ┆ 1976    ┆ "CTGTATTTCCGCTAGA-1" ┆ 2             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1955    ┆ 1976    ┆ "CTTTAGTTCCACCCTG-1" ┆ 3             │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ "JH584295.1" ┆ 1955    ┆ 1976    ┆ "GACCTAGTCCTCAGCT-1" ┆ 1             │
└──────────────┴─────────┴─────────┴──────────────────────┴───────────────┘
```

**Comment by ritchie46** [comment]
> polars categorical to arrow dict is very fast now (zero copy I assume).

Almost. Polars uses a `u32` for the keys. Pyarrow does not allow unsigned integers as keys, so they coerce that to `i32`. I believe pyarrow also casts the indices of the list data from `i64` to `i32`. So this is half copy.

> arrow dict to polars categorical requires some time (but works).

This is zero copy in FFI. But then the polars categorical hashmap needs to be build and that requires work.

**Comment by jorgecarleitao** [comment]
> Pyarrow does not allow unsigned integers as keys, so they coerce that to `i32`.

That is a bug on pyarrow: C++ supports both signed and unsigned keys in dictionaries. Do you have a small repro that we can push to Arrow's JIRA?

**Comment by ritchie46** [comment]
The first hint I encountered was this:

```python
>>> import pyarrow as pa
>>> pa.dictionary(pa.uint32(), pa.utf8())
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_30593/840774823.py in <module>
----> 1 pa.dictionary(pa.uint32(), pa.utf8())

/opt/miniconda3/lib/python3.9/site-packages/pyarrow/types.pxi in pyarrow.lib.dictionary()

/opt/miniconda3/lib/python3.9/site-packages/pyarrow/types.pxi in pyarrow.lib.dictionary()

TypeError: The dictionary index type should be signed integer.
```

The second is was sending a `Dictionary<u32, large-utf8>` over FFI made the type in pyarrow a `Dictionary<i32, utf8>`. Shall I wrap a snippet for that?

**Comment by jorgecarleitao** [comment]
Thanks! Filed: https://issues.apache.org/jira/browse/ARROW-14043.

**Comment by ghuls** [comment]
Seems to be fixed in arrow master: https://github.com/apache/arrow/commit/6c23e1765df2117ad481c53cca3e6b177b5b4e9a

**Comment by ritchie46** [comment]
> Seems to be fixed in arrow master: [apache/arrow@6c23e17](https://github.com/apache/arrow/commit/6c23e1765df2117ad481c53cca3e6b177b5b4e9a)

Nice, that was fast.
